File: py-packages/qgis-sdk/src/qgis_sdk/bridge/qgis_api/project.py
"""qgis_sdk.bridge.qgis_api.project — project API."""

from __future__ import annotations

from typing import Any, Dict


def _get_project():
    try:
        from qgis.core import QgsProject
        return QgsProject.instance()
    except ImportError:
        return None


class ProjectAPI:
    def info(self) -> Dict[str, Any]:
        project = _get_project()
        if not project:
            return {"path": "", "crs": "", "title": ""}
        try:
            return {
                "path": project.fileName(),
                "crs": project.crs().authid() if hasattr(project, "crs") else "",
                "title": project.title() if hasattr(project, "title") else "",
            }
        except Exception as e:
            return {"path": "", "crs": "", "title": "", "error": str(e)}

    def write(self) -> bool:
        project = _get_project()
        if not project:
            return False
        try:
            return project.write()
        except Exception:
            return False

    def crs(self) -> str:
        project = _get_project()
        if not project:
            return ""
        try:
            return project.crs().authid()
        except Exception:
            return ""

    def set_crs(self, authid: str) -> bool:
        project = _get_project()
        if not project:
            return False
        try:
            from qgis.core import QgsCoordinateReferenceSystem
            crs = QgsCoordinateReferenceSystem(authid)
            project.setCrs(crs)
            return True
        except Exception as e:
            print(f"[ProjectAPI] set_crs failed: {e}")
            return False

    def path(self) -> str:
        project = _get_project()
        if not project:
            return ""
        try:
            return project.fileName()
        except Exception:
            return ""

File: ts-packages/qgis-sdk-bridge/src/qgis/project.ts
/** qgis.project API */

import type { QgisBridge } from '../window';

export interface QgisProjectInfo {
  path: string;
  crs: string;
  title: string;
}

export class ProjectAPI {
  constructor(private _bridge: QgisBridge, private _raw: any) {}

  private async _call(method: string, ...args: any[]): Promise<any> {
    if (this._raw && typeof this._raw[method] === 'function') {
      return new Promise((resolve, reject) => {
        try {
          this._raw[method](...args, (res: any) => {
            if (typeof res === 'string') {
              try {
                const parsed = JSON.parse(res);
                resolve(parsed);
                return;
              } catch {}
            }
            resolve(res);
          });
        } catch (e) {
          reject(e);
        }
      });
    }
    if (this._bridge && typeof (this._bridge as any)[method] === 'function') {
      return (this._bridge as any)[method](...args);
    }
    console.warn(`[qgis.project] ${method} mock`);
    if (method === 'project_info') return { path: '', crs: '', title: '' };
    if (method === 'project_crs') return '';
    if (method === 'project_path') return '';
    return true;
  }

  async info(): Promise<QgisProjectInfo> {
    return this._call('project_info');
  }

  async write(): Promise<boolean> {
    return this._call('project_write');
  }

  async crs(): Promise<string> {
    return this._call('project_crs');
  }

  async setCrs(authid: string): Promise<boolean> {
    return this._call('project_set_crs', authid);
  }

  async path(): Promise<string> {
    return this._call('project_path');
  }
}
